module/zfs/dbuf.c

```diff
@@ -3480,12 +3488,12 @@ dbuf_check_crypt(dbuf_dirty_record_t *dr)
 		    &zb, B_TRUE);
 		if (err)
 			panic("Invalid dnode block MAC");
-	} else if (dr->dt.dl.dr_raw) {
-		/*
-		 * Writing raw encrypted data requires the db's arc buffer
-		 * to be converted to raw by the caller.
-		 */
-		ASSERT(arc_is_encrypted(db->db_buf));
+	} else if (dr->dt.dl.dr_has_raw_params) {
+		(void) arc_release(dr->dt.dl.dr_data, db);
+		arc_convert_to_raw(dr->dt.dl.dr_data,
+		    dmu_objset_id(db->db_objset),
+		    dr->dt.dl.dr_byteorder, DMU_OT_DNODE,
+		    dr->dt.dl.dr_salt, dr->dt.dl.dr_iv, dr->dt.dl.dr_mac);
```

Comment:
By "other kinds of raw writes", do you mean arc_loan_raw_buf()?  I'm not sure how to use that knowledge to change this function name.  Did you have a new name in mind?

Reply:
> By "other kinds of raw writes", do you mean arc_loan_raw_buf()?

Yes.

> Did you have a new name in mind?

Nothing specific. Perhaps something like `dmu_buf_set_dnode_crypt_params()`? Or perhaps just an ASSERT at the beginning of the function making sure we only call this function with L0 metadnode blocks. Just because otherwise the parameters just sit ignored in the dirty record.

Reply:
Ah, I see.  Yes, we should totally add an assertion here that it's only for L0 dnode blocks.